### ISSUE TITLE
Will log_determinant return inf for singular matrices?

### ISSUE THREAD

**Issue by xlvector** [deepmind/ferminet#5]
In the last line of following code

`
def log_determinant(x):
  with tf.name_scope('determinant'):
    with tf.device('/cpu:0'):
      s, u, v = tf.linalg.svd(x)

    sign = sign_det(x)

    def grad(dsign, dlog):
      del dsign
      # d ln|det(X)|/dX = transpose(X^{-1}). This uses pseudo-inverse via SVD
      # and so is not numerically stable...
      adj = tf.matmul(u,
                      tf.matmul(tf.linalg.diag(1.0 / s), v, transpose_b=True))
      return tf.expand_dims(tf.expand_dims(dlog, -1), -1) * adj

    return (sign, tf.reduce_sum(tf.log(s), axis=-1)), grad
`

it seems that ,for singular matrices, s will have zero elements and tf.log(s) will return -inf ?
And tf.linalg.diag(1.0 / s) many also generate inf results?

**Comment by dpfau** [comment]
That would be correct behavior. More importantly, logdet_matmul will *not* return -inf, nor will inf or nan appear in the gradients, as long as at least one of the matrices in the batch is non-singular. So for multi-determinant wavefunctions, it's extremely unlikely to blow up due to one of the inputs being singular.
